packages/fpdart/lib/src/effect.dart

```diff
@@ -0,0 +1,561 @@
+        Left(value: final cause) => throw cause,
+        Right(value: final value) => value,
+      };
+    } on Cause<L> {
+      rethrow;
+    } catch (error, stackTrace) {
+      throw Die(error, stackTrace);
+    }
+  }
+
+  /// {@category execution}
+  Future<Exit<L, R>> runFutureExit(E env) async {
+    try {
+      final result = _unsafeRun(env);
+      if (result is! Future) {
+        return result;
+      }
+      return result;
+    } on Cause<L> catch (cause) {
+      return Left(cause);
+    } catch (error, stackTrace) {
+      return Left(Die(error, stackTrace));
+    }
+  }
+
+  /// {@category constructors}
+  factory Effect.gen(DoFunctionEffect<E, L, R> f) => Effect<E, L, R>._(
+        (env) {
+          try {
+            return f(_effectGen<E, L>(env)).then(
+              Right.new,
+              onError: (error) {
+                if (error is _EffectThrow<L>) {
+                  return Left<Cause<L>, R>(error.cause);
+                }
+
+                return Left<Cause<L>, R>(Die.current(error));
+              },
+            );
+          } on _EffectThrow<L> catch (genError) {
+            return Left(genError.cause);
+          }
+        },
+      );
+
+  /// {@category constructors}
+  factory Effect.tryCatch({
+    required FutureOr<R> Function() execute,
+    required L Function(Object error, StackTrace stackTrace) onError,
+    FutureOr<dynamic> Function()? onCancel,
+  }) =>
+      Effect._(
+        (env) {
+          try {
+            return execute().then(Right.new);
+          } catch (err, stack) {
+            return Left(Fail(onError(err, stack), stack));
+          }
+        },
+      );
+
+  /// {@category constructors}
+  factory Effect.fromNullable(R? value, {required L Function() onNull}) =>
+      Effect._(
+        (_) => value == null ? Left(Fail(onNull())) : Right(value),
+      );
+
+  /// {@category constructors}
+  factory Effect.functionFail(FutureOr<Cause<L>> Function() f) => Effect._(
+        (_) => f().then(Left.new),
+      );
+
+  /// {@category constructors}
+  factory Effect.functionSucceed(FutureOr<R> Function() f) => Effect._(
+        (_) => f().then(Right.new),
+      );
+
+  /// {@category constructors}
+  factory Effect.fail(L value) => Effect._((_) => Left(Fail(value)));
+
+  /// {@category constructors}
+  factory Effect.succeed(R value) => Effect._((_) => Right(value));
+
+  /// {@category constructors}
+  static Effect<E, Never, Never> die<E>(dynamic defect) => Effect._(
+        (_) => Left(Die.current(defect)),
+      );
+
+  /// {@category constructors}
+  static Effect<E, Never, Never> functionDie<E>(dynamic Function() run) =>
+      Effect._(
+        (_) => Left(Die.current(run())),
+      );
+
+  /// {@category constructors}
+  static Effect<E, L, fpdart_unit.Unit> unit<E, L>() => Effect._(
+        (_) => const Right(fpdart_unit.unit),
+      );
+
+  /// {@category collecting}
+  static Effect<E, L, Iterable<R>> forEach<E, L, R, A>(
```

Comment:
Concurrency?

Reply:
Working on it, I am thinking how to make it work well with `Iterable` (not just `List`)